exorcist/taskdb.py

```diff
@@ -1,40 +1,226 @@
+import abc
 import sqlalchemy as sqla
 
+# remaining imports are for typing
+from typing import Optional, Iterable
+from .models import TaskStatus, Task
+from os import PathLike
+import networkx as nx
 
-def create_task_db(engine):
+
+class NoStatusChange(Exception):
+    """Raised when an attempt to change task status does not change status.
+
+    In most cases, like when a task status is changed to "completed", if the
+    task is already marked as the new status, this indicates database
+    corruption.
+
+    Raising this exception has two purposes:
+
+    1. To enable it to be caught and allow the database corruption to be
+       fixed, if possible, and exit otherwise.
+    2. So that, in the event that the status that needs to be set on early
+       exit (e.g., due to a sigterm send by the queuing system) is the same
+       as the current status, it can be caught without error and exit
+       normally.
     """
-    Create an empty task database with the provided engine.
+
+
+class AbstractTaskStatusDB(abc.ABC):
+    @abc.abstractmethod
+    def add_task(self, taskid: str, requirements: Iterable[str]):
+        """Add a task to the database.
+
+        Parameters
+        ----------
+        taskid: str
+            the taskid to add to the database
+        requirements: Iterable[str]
+            taskids that directly block the task to be added (typically,
+            whose outputs are inputs to the task)
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def add_task_network(self, task_network: nx.DiGraph):
+        """Add a network of tasks to the database.
+
+        Parameters
+        ----------
+        task_network: nx.Digraph
+            A network with taskid strings as nodes.
```

Comment:
if we're doing this it's worth documenting which way round the directional edges work.  Do nodes point to children or parents?

Reply:
Part of #14. This is just supposed to be a rough skeleton here; let's worry about this when we implement these things.